Broken metrics server on seed with cluster namespace

**User Story**
The metrics server was moved from user cluster to seed with cluster namespace: https://github.com/kubermatic/kubermatic/commit/90f41f477fa20e6df08f0057cb05434783983f13

 The metrics server doesn't work right now. 
The API is fixed to use proper metrics server but also was broken: https://github.com/kubermatic/kubermatic/pull/5480

```
kubectl logs metrics-server-848b6cddc9-59jkq  -n cluster-f7xd8qc7dk metrics-server
{"level":"info","time":"2020-06-05T06:54:05.365Z","logger":"http-prober","caller":"http-prober/main.go:109","msg":"Probing","attempt":1,"max-attempts":100,"target":"https://apiserver-external.cluster-f7xd8qc7dk.svc.cluster.local.:30331/healthz"}
{"level":"info","time":"2020-06-05T06:54:05.374Z","logger":"http-prober","caller":"http-prober/main.go:98","msg":"Hostname resolved","hostname":"apiserver-external.cluster-f7xd8qc7dk.svc.cluster.local.","address":"10.47.255.162:30331"}
{"level":"info","time":"2020-06-05T06:54:05.376Z","logger":"http-prober","caller":"http-prober/main.go:122","msg":"Endpoint is available"}
W0605 06:54:07.966421       1 authentication.go:262] Unable to get configmap/extension-apiserver-authentication in kube-system.  Usually fixed by 'kubectl create rolebinding -n kube-system ROLEBINDING_NAME --role=extension-apiserver-authentication-reader --serviceaccount=YOUR_NS:YOUR_SA'
Error: configmaps "extension-apiserver-authentication" is forbidden: User "metrics-server" cannot get resource "configmaps" in API group "" in the namespace "kube-system"
Usage:
   [flags]

Flags:
      --alsologtostderr                                         log to standard error as well as files
      --authentication-kubeconfig string                        kubeconfig file pointing at the 'core' kubernetes server with enough rights to create tokenaccessreviews.authentication.k8s.io.
      --authentication-skip-lookup                              If false, the authentication-kubeconfig will be used to lookup missing authentication configuration from the cluster.
      --authentication-token-webhook-cache-ttl duration         The duration to cache responses from the webhook token authenticator. (default 10s)
      --authentication-tolerate-lookup-failure                  If true, failures to look up missing authentication configuration from the cluster are not considered fatal. Note that this can result in authentication that treats all requests as anonymous.
      --authorization-always-allow-paths strings                A list of HTTP paths to skip during authorization, i.e. these are authorized without contacting the 'core' kubernetes server.
      --authorization-kubeconfig string                         kubeconfig file pointing at the 'core' kubernetes server with enough rights to create subjectaccessreviews.authorization.k8s.io.
      --authorization-webhook-cache-authorized-ttl duration     The duration to cache 'authorized' responses from the webhook authorizer. (default 10s)
      --authorization-webhook-cache-unauthorized-ttl duration   The duration to cache 'unauthorized' responses from the webhook authorizer. (default 10s)
      --bind-address ip                                         The IP address on which to listen for the --secure-port port. The associated interface(s) must be reachable by the rest of the cluster, and by CLI/web clients. If blank, all interfaces will be used (0.0.0.0 for all IPv4 interfaces and :: for all IPv6 interfaces). (default 0.0.0.0)
      --cert-dir string                                         The directory where the TLS certs are located. If --tls-cert-file and --tls-private-key-file are provided, this flag will be ignored. (default "apiserver.local.config/certificates")
      --client-ca-file string                                   If set, any request presenting a client certificate signed by one of the authorities in the client-ca-file is authenticated with an identity corresponding to the CommonName of the client certificate.
      --contention-profiling                                    Enable lock contention profiling, if profiling is enabled
  -h, --help                                                    help for this command
      --http2-max-streams-per-connection int                    The limit that the server gives to clients for the maximum number of streams in an HTTP/2 connection. Zero means to use golang's default.
      --kubeconfig string                                       The path to the kubeconfig used to connect to the Kubernetes API server and the Kubelets (defaults to in-cluster config)
      --kubelet-certificate-authority string                    Path to the CA to use to validate the Kubelet's serving certificates.
      --kubelet-insecure-tls                                    Do not verify CA of serving certificates presented by Kubelets.  For testing purposes only.
      --kubelet-port int                                        The port to use to connect to Kubelets. (default 10250)
      --kubelet-preferred-address-types strings                 The priority of node address types to use when determining which address to use to connect to a particular node (default [Hostname,InternalDNS,InternalIP,ExternalDNS,ExternalIP])
      --log-flush-frequency duration                            Maximum number of seconds between log flushes (default 5s)
      --log_backtrace_at traceLocation                          when logging hits line file:N, emit a stack trace (default :0)
      --log_dir string                                          If non-empty, write log files in this directory
      --log_file string                                         If non-empty, use this log file
      --logtostderr                                             log to standard error instead of files (default true)
      --metric-resolution duration                              The resolution at which metrics-server will retain metrics. (default 1m0s)
      --profiling                                               Enable profiling via web interface host:port/debug/pprof/ (default true)
      --requestheader-allowed-names strings                     List of client certificate common names to allow to provide usernames in headers specified by --requestheader-username-headers. If empty, any client certificate validated by the authorities in --requestheader-client-ca-file is allowed.
      --requestheader-client-ca-file string                     Root certificate bundle to use to verify client certificates on incoming requests before trusting usernames in headers specified by --requestheader-username-headers. WARNING: generally do not depend on authorization being already done for incoming requests.
      --requestheader-extra-headers-prefix strings              List of request header prefixes to inspect. X-Remote-Extra- is suggested. (default [x-remote-extra-])
      --requestheader-group-headers strings                     List of request headers to inspect for groups. X-Remote-Group is suggested. (default [x-remote-group])
      --requestheader-username-headers strings                  List of request headers to inspect for usernames. X-Remote-User is common. (default [x-remote-user])
      --secure-port int                                         The port on which to serve HTTPS with authentication and authorization.If 0, don't serve HTTPS at all. (default 443)
      --skip_headers                                            If true, avoid header prefixes in the log messages
      --stderrthreshold severity                                logs at or above this threshold go to stderr
      --tls-cert-file string                                    File containing the default x509 Certificate for HTTPS. (CA cert, if any, concatenated after server cert). If HTTPS serving is enabled, and --tls-cert-file and --tls-private-key-file are not provided, a self-signed certificate and key are generated for the public address and saved to the directory specified by --cert-dir.
      --tls-cipher-suites strings                               Comma-separated list of cipher suites for the server. If omitted, the default Go cipher suites will be use.  Possible values: TLS_ECDHE_ECDSA_WITH_AES_128_CBC_SHA,TLS_ECDHE_ECDSA_WITH_AES_128_CBC_SHA256,TLS_ECDHE_ECDSA_WITH_AES_128_GCM_SHA256,TLS_ECDHE_ECDSA_WITH_AES_256_CBC_SHA,TLS_ECDHE_ECDSA_WITH_AES_256_GCM_SHA384,TLS_ECDHE_ECDSA_WITH_CHACHA20_POLY1305,TLS_ECDHE_ECDSA_WITH_RC4_128_SHA,TLS_ECDHE_RSA_WITH_3DES_EDE_CBC_SHA,TLS_ECDHE_RSA_WITH_AES_128_CBC_SHA,TLS_ECDHE_RSA_WITH_AES_128_CBC_SHA256,TLS_ECDHE_RSA_WITH_AES_128_GCM_SHA256,TLS_ECDHE_RSA_WITH_AES_256_CBC_SHA,TLS_ECDHE_RSA_WITH_AES_256_GCM_SHA384,TLS_ECDHE_RSA_WITH_CHACHA20_POLY1305,TLS_ECDHE_RSA_WITH_RC4_128_SHA,TLS_RSA_WITH_3DES_EDE_CBC_SHA,TLS_RSA_WITH_AES_128_CBC_SHA,TLS_RSA_WITH_AES_128_CBC_SHA256,TLS_RSA_WITH_AES_128_GCM_SHA256,TLS_RSA_WITH_AES_256_CBC_SHA,TLS_RSA_WITH_AES_256_GCM_SHA384,TLS_RSA_WITH_RC4_128_SHA
      --tls-min-version string                                  Minimum TLS version supported. Possible values: VersionTLS10, VersionTLS11, VersionTLS12
      --tls-private-key-file string                             File containing the default x509 private key matching --tls-cert-file.
      --tls-sni-cert-key namedCertKey                           A pair of x509 certificate and private key file paths, optionally suffixed with a list of domain patterns which are fully qualified domain names, possibly with prefixed wildcard segments. If no domain patterns are provided, the names of the certificate are extracted. Non-wildcard matches trump over wildcard matches, explicit domain patterns trump over extracted names. For multiple key/certificate pairs, use the --tls-sni-cert-key multiple times. Examples: "example.crt,example.key" or "foo.crt,foo.key:*.foo.com,foo.com". (default [])
  -v, --v Level                                                 number for the log level verbosity (default 0)
      --vmodule moduleSpec                                      comma-separated list of pattern=N settings for file-filtered logging

panic: configmaps "extension-apiserver-authentication" is forbidden: User "metrics-server" cannot get resource "configmaps" in API group "" in the namespace "kube-system"

goroutine 1 [running]:
main.main()
	/go/src/github.com/kubernetes-incubator/metrics-server/cmd/metrics-server/metrics-server.go:39 +0x13b

```

--- a/addons/metrics-server/auth-delegator.yaml
+++ b/addons/metrics-server/auth-delegator.yaml
@@ -1,0 +1,13 @@
+---
+apiVersion: rbac.authorization.k8s.io/v1beta1
+kind: ClusterRoleBinding
+metadata:
+  name: metrics-server:system:auth-delegator
+roleRef:
+  apiGroup: rbac.authorization.k8s.io
+  kind: ClusterRole
+  name: system:auth-delegator
+subjects:
+- kind: ServiceAccount
+  name: metrics-server
+  namespace: kube-system

--- a/addons/metrics-server/auth-reader.yaml
+++ b/addons/metrics-server/auth-reader.yaml
@@ -1,0 +1,14 @@
+---
+apiVersion: rbac.authorization.k8s.io/v1beta1
+kind: RoleBinding
+metadata:
+  name: metrics-server-auth-reader
+  namespace: kube-system
+roleRef:
+  apiGroup: rbac.authorization.k8s.io
+  kind: Role
+  name: extension-apiserver-authentication-reader
+subjects:
+- kind: ServiceAccount
+  name: metrics-server
+  namespace: kube-system

--- a/addons/metrics-server/metrics-apiservice.yaml
+++ b/addons/metrics-server/metrics-apiservice.yaml
@@ -1,0 +1,14 @@
+---
+apiVersion: apiregistration.k8s.io/v1beta1
+kind: APIService
+metadata:
+  name: v1beta1.metrics.k8s.io
+spec:
+  service:
+    name: metrics-server
+    namespace: kube-system
+  group: metrics.k8s.io
+  version: v1beta1
+  insecureSkipTLSVerify: true
+  groupPriorityMinimum: 100
+  versionPriority: 100

--- a/addons/metrics-server/metrics-server-deployment.yaml
+++ b/addons/metrics-server/metrics-server-deployment.yaml
@@ -1,0 +1,40 @@
+---
+apiVersion: v1
+kind: ServiceAccount
+metadata:
+  name: metrics-server
+  namespace: kube-system
+---
+apiVersion: apps/v1
+kind: Deployment
+metadata:
+  name: metrics-server
+  namespace: kube-system
+  labels:
+    k8s-app: metrics-server
+spec:
+  selector:
+    matchLabels:
+      k8s-app: metrics-server
+  template:
+    metadata:
+      name: metrics-server
+      labels:
+        k8s-app: metrics-server
+    spec:
+      serviceAccountName: metrics-server
+      volumes:
+      # mount in tmp so we can safely use from-scratch images and/or read-only containers
+      - name: tmp-dir
+        emptyDir: {}
+      containers:
+      - name: metrics-server
+        image: '{{ Registry "gcr.io" }}/google_containers/metrics-server-amd64:v0.2.1'
+        command:
+        - /metrics-server
+        - '--source=kubernetes.summary_api:https://kubernetes.default.svc?kubeletHttps=true&kubeletPort=10250&insecure=true'
+        imagePullPolicy: Always
+        volumeMounts:
+        - name: tmp-dir
+          mountPath: /tmp
+

--- a/addons/metrics-server/metrics-server-service.yaml
+++ b/addons/metrics-server/metrics-server-service.yaml
@@ -1,0 +1,15 @@
+---
+apiVersion: v1
+kind: Service
+metadata:
+  name: metrics-server
+  namespace: kube-system
+  labels:
+    kubernetes.io/name: "Metrics-server"
+spec:
+  selector:
+    k8s-app: metrics-server
+  ports:
+  - port: 443
+    protocol: TCP
+    targetPort: 443

--- a/addons/metrics-server/resource-reader.yaml
+++ b/addons/metrics-server/resource-reader.yaml
@@ -1,0 +1,38 @@
+---
+apiVersion: rbac.authorization.k8s.io/v1
+kind: ClusterRole
+metadata:
+  name: system:metrics-server
+rules:
+- apiGroups:
+  - ""
+  resources:
+  - pods
+  - nodes
+  - nodes/stats
+  - namespaces
+  verbs:
+  - get
+  - list
+  - watch
+- apiGroups:
+  - "extensions"
+  resources:
+  - deployments
+  verbs:
+  - get
+  - list
+  - watch
+---
+apiVersion: rbac.authorization.k8s.io/v1
+kind: ClusterRoleBinding
+metadata:
+  name: system:metrics-server
+roleRef:
+  apiGroup: rbac.authorization.k8s.io
+  kind: ClusterRole
+  name: system:metrics-server
+subjects:
+- kind: ServiceAccount
+  name: metrics-server
+  namespace: kube-system

--- a/api/cmd/seed-controller-manager/options.go
+++ b/api/cmd/seed-controller-manager/options.go
@@ -225,6 +225,14 @@ func (o controllerRunOptions) validate() error {
 		return fmt.Errorf("failed to parse nodePortRange: %v", err)
 	}
 
+	// Validate the metrics-server addon is disabled, otherwise it creates conflicts with the resources
+	// we create for the metrics-server running in the seed and will render the latter unusable
+	for _, addon := range o.kubernetesAddons.Items {
+		if addon.Name == "metrics-server" {
+			return errors.New("the metrics-server addon must be disabled, it is now deployed inside the seed cluster")
+		}
+	}
+
 	return nil
 }
 

--- a/api/pkg/controller/user-cluster-controller-manager/resources/reconciler.go
+++ b/api/pkg/controller/user-cluster-controller-manager/resources/reconciler.go
@@ -13,6 +13,7 @@ import (
 	kubestatemetrics "github.com/kubermatic/kubermatic/api/pkg/controller/user-cluster-controller-manager/resources/resources/kube-state-metrics"
 	kubernetesdashboard "github.com/kubermatic/kubermatic/api/pkg/controller/user-cluster-controller-manager/resources/resources/kubernetes-dashboard"
 	machinecontroller "github.com/kubermatic/kubermatic/api/pkg/controller/user-cluster-controller-manager/resources/resources/machine-controller"
+	metricsserver "github.com/kubermatic/kubermatic/api/pkg/controller/user-cluster-controller-manager/resources/resources/metrics-server"
 	nodelocaldns "github.com/kubermatic/kubermatic/api/pkg/controller/user-cluster-controller-manager/resources/resources/node-local-dns"
 	"github.com/kubermatic/kubermatic/api/pkg/controller/user-cluster-controller-manager/resources/resources/openshift"
 	"github.com/kubermatic/kubermatic/api/pkg/controller/user-cluster-controller-manager/resources/resources/openvpn"
@@ -126,7 +127,9 @@ func (r *reconciler) reconcile(ctx context.Context) error {
 
 func (r *reconciler) ensureAPIServices(ctx context.Context, data reconcileData) error {
 	caCert := triple.EncodeCertPEM(data.caCert.Cert)
-	creators := []reconciling.NamedAPIServiceCreatorGetter{}
+	creators := []reconciling.NamedAPIServiceCreatorGetter{
+		metricsserver.APIServiceCreator(caCert),
+	}
 
 	if r.openshift {
 		openshiftAPIServiceCreators, err := openshift.GetAPIServicesForOpenshiftVersion(r.version, caCert)
@@ -245,6 +248,7 @@ func (r *reconciler) reconcileRoleBindings(ctx context.Context) error {
 	// kube-system
 	creators := []reconciling.NamedRoleBindingCreatorGetter{
 		machinecontroller.KubeSystemRoleBindingCreator(),
+		metricsserver.RolebindingAuthReaderCreator(),
 		scheduler.RoleBindingAuthDelegator(),
 		controllermanager.RoleBindingAuthDelegator(),
 		clusterautoscaler.KubeSystemRoleBindingCreator(),
@@ -310,6 +314,7 @@ func (r *reconciler) reconcileClusterRoles(ctx context.Context) error {
 		prometheus.ClusterRoleCreator(),
 		machinecontroller.ClusterRoleCreator(),
 		dnatcontroller.ClusterRoleCreator(),
+		metricsserver.ClusterRoleCreator(),
 		clusterautoscaler.ClusterRoleCreator(),
 	}
 
@@ -336,6 +341,8 @@ func (r *reconciler) reconcileClusterRoleBindings(ctx context.Context) error {
 		machinecontroller.NodeBootstrapperClusterRoleBindingCreator(),
 		machinecontroller.NodeSignerClusterRoleBindingCreator(),
 		dnatcontroller.ClusterRoleBindingCreator(),
+		metricsserver.ClusterRoleBindingResourceReaderCreator(),
+		metricsserver.ClusterRoleBindingAuthDelegatorCreator(),
 		scheduler.ClusterRoleBindingAuthDelegatorCreator(),
 		controllermanager.ClusterRoleBindingAuthDelegator(),
 		clusterautoscaler.ClusterRoleBindingCreator(),
@@ -388,7 +395,9 @@ func (r *reconciler) reconcileMutatingWebhookConfigurations(
 }
 
 func (r *reconciler) reconcileServices(ctx context.Context) error {
-	creatorsKubeSystem := []reconciling.NamedServiceCreatorGetter{}
+	creatorsKubeSystem := []reconciling.NamedServiceCreatorGetter{
+		metricsserver.ExternalNameServiceCreator(r.namespace),
+	}
 
 	if err := reconciling.ReconcileServices(ctx, creatorsKubeSystem, metav1.NamespaceSystem, r.Client); err != nil {
 		return fmt.Errorf("failed to reconcile Services in kube-system namespace: %v", err)

--- a/api/pkg/controller/user-cluster-controller-manager/resources/resources/metrics-server/apiservice.go
+++ b/api/pkg/controller/user-cluster-controller-manager/resources/resources/metrics-server/apiservice.go
@@ -1,0 +1,37 @@
+package metricsserver
+
+import (
+	"github.com/kubermatic/kubermatic/api/pkg/resources"
+	"github.com/kubermatic/kubermatic/api/pkg/resources/reconciling"
+	apiregistrationv1beta1 "k8s.io/kube-aggregator/pkg/apis/apiregistration/v1beta1"
+
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+const (
+	Name = "metrics-server"
+)
+
+// APIServiceCreator returns the func to create/update the APIService used by the metrics-server
+func APIServiceCreator(caBundle []byte) reconciling.NamedAPIServiceCreatorGetter {
+	return func() (string, reconciling.APIServiceCreator) {
+		return resources.MetricsServerAPIServiceName, func(se *apiregistrationv1beta1.APIService) (*apiregistrationv1beta1.APIService, error) {
+			labels := resources.BaseAppLabels(Name, nil)
+			se.Labels = labels
+
+			if se.Spec.Service == nil {
+				se.Spec.Service = &apiregistrationv1beta1.ServiceReference{}
+			}
+			se.Spec.Service.Namespace = metav1.NamespaceSystem
+			se.Spec.Service.Name = resources.MetricsServerExternalNameServiceName
+			se.Spec.Group = "metrics.k8s.io"
+			se.Spec.Version = "v1beta1"
+			se.Spec.InsecureSkipTLSVerify = false
+			se.Spec.CABundle = caBundle
+			se.Spec.GroupPriorityMinimum = 100
+			se.Spec.VersionPriority = 100
+
+			return se, nil
+		}
+	}
+}

--- a/api/pkg/controller/user-cluster-controller-manager/resources/resources/metrics-server/clusterrole.go
+++ b/api/pkg/controller/user-cluster-controller-manager/resources/resources/metrics-server/clusterrole.go
@@ -1,0 +1,42 @@
+package metricsserver
+
+import (
+	"github.com/kubermatic/kubermatic/api/pkg/resources"
+	"github.com/kubermatic/kubermatic/api/pkg/resources/reconciling"
+
+	rbacv1 "k8s.io/api/rbac/v1"
+)
+
+// ClusterRole returns a cluster role for the metrics server
+func ClusterRoleCreator() reconciling.NamedClusterRoleCreatorGetter {
+	return func() (string, reconciling.ClusterRoleCreator) {
+		return resources.MetricsServerClusterRoleName, func(cr *rbacv1.ClusterRole) (*rbacv1.ClusterRole, error) {
+			cr.Labels = resources.BaseAppLabels(Name, nil)
+
+			cr.Rules = []rbacv1.PolicyRule{
+				{
+					APIGroups: []string{""},
+					Resources: []string{
+						"pods",
+						"nodes",
+						"nodes/stats",
+						"namespaces",
+					},
+					Verbs: []string{
+						"get",
+						"list",
+						"watch",
+					},
+				},
+				{
+					APIGroups: []string{"extensions"},
+					Resources: []string{
+						"deployments",
+					},
+					Verbs: []string{"get", "list", "watch"},
+				},
+			}
+			return cr, nil
+		}
+	}
+}

--- a/api/pkg/controller/user-cluster-controller-manager/resources/resources/metrics-server/clusterrolebinding.go
+++ b/api/pkg/controller/user-cluster-controller-manager/resources/resources/metrics-server/clusterrolebinding.go
@@ -1,0 +1,37 @@
+package metricsserver
+
+import (
+	"github.com/kubermatic/kubermatic/api/pkg/resources"
+	"github.com/kubermatic/kubermatic/api/pkg/resources/reconciling"
+
+	rbacv1 "k8s.io/api/rbac/v1"
+)
+
+// ClusterRoleBindingResourceReaderCreator returns the ClusterRoleBinding required for the metrics server to read all required resources
+func ClusterRoleBindingResourceReaderCreator() reconciling.NamedClusterRoleBindingCreatorGetter {
+	return func() (string, reconciling.ClusterRoleBindingCreator) {
+		return resources.MetricsServerResourceReaderClusterRoleBindingName, func(crb *rbacv1.ClusterRoleBinding) (*rbacv1.ClusterRoleBinding, error) {
+			crb.Labels = resources.BaseAppLabels(Name, nil)
+
+			crb.RoleRef = rbacv1.RoleRef{
+				Name:     resources.MetricsServerClusterRoleName,
+				Kind:     "ClusterRole",
+				APIGroup: rbacv1.GroupName,
+			}
+			crb.Subjects = []rbacv1.Subject{
+				{
+					Kind:     "User",
+					Name:     resources.MetricsServerCertUsername,
+					APIGroup: rbacv1.GroupName,
+				},
+			}
+			return crb, nil
+		}
+	}
+
+}
+
+// ClusterRoleBindingAuthDelegatorCreator returns the ClusterRoleBinding required for the metrics server to create token review requests
+func ClusterRoleBindingAuthDelegatorCreator() reconciling.NamedClusterRoleBindingCreatorGetter {
+	return resources.ClusterRoleBindingAuthDelegatorCreator(resources.MetricsServerCertUsername)
+}

--- a/api/pkg/controller/user-cluster-controller-manager/resources/resources/metrics-server/external-name-service.go
+++ b/api/pkg/controller/user-cluster-controller-manager/resources/resources/metrics-server/external-name-service.go
@@ -1,0 +1,26 @@
+package metricsserver
+
+import (
+	"fmt"
+
+	"github.com/kubermatic/kubermatic/api/pkg/resources"
+	"github.com/kubermatic/kubermatic/api/pkg/resources/reconciling"
+
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+// ExternalNameServiceCreator returns the function to reconcile the metrics server service
+func ExternalNameServiceCreator(namespace string) reconciling.NamedServiceCreatorGetter {
+	return func() (string, reconciling.ServiceCreator) {
+		return resources.MetricsServerExternalNameServiceName, func(se *corev1.Service) (*corev1.Service, error) {
+			se.Namespace = metav1.NamespaceSystem
+			se.Labels = resources.BaseAppLabels(Name, nil)
+
+			se.Spec.Type = corev1.ServiceTypeExternalName
+			se.Spec.ExternalName = fmt.Sprintf("%s.%s.svc.cluster.local", resources.MetricsServerServiceName, namespace)
+
+			return se, nil
+		}
+	}
+}

--- a/api/pkg/controller/user-cluster-controller-manager/resources/resources/metrics-server/rolebinding.go
+++ b/api/pkg/controller/user-cluster-controller-manager/resources/resources/metrics-server/rolebinding.go
@@ -1,0 +1,31 @@
+package metricsserver
+
+import (
+	"github.com/kubermatic/kubermatic/api/pkg/resources"
+	"github.com/kubermatic/kubermatic/api/pkg/resources/reconciling"
+
+	rbacv1 "k8s.io/api/rbac/v1"
+)
+
+// RolebindingAuthReaderCreator returns a func to create/update the RoleBinding used by the metrics-server to get access to the token subject review API
+func RolebindingAuthReaderCreator() reconciling.NamedRoleBindingCreatorGetter {
+	return func() (string, reconciling.RoleBindingCreator) {
+		return resources.MetricsServerAuthReaderRoleName, func(rb *rbacv1.RoleBinding) (*rbacv1.RoleBinding, error) {
+			rb.Labels = resources.BaseAppLabels(Name, nil)
+
+			rb.RoleRef = rbacv1.RoleRef{
+				Name:     "extension-apiserver-authentication-reader",
+				Kind:     "Role",
+				APIGroup: rbacv1.GroupName,
+			}
+			rb.Subjects = []rbacv1.Subject{
+				{
+					Kind:     "User",
+					Name:     resources.MetricsServerCertUsername,
+					APIGroup: rbacv1.GroupName,
+				},
+			}
+			return rb, nil
+		}
+	}
+}

--- a/api/pkg/handler/v1/cluster/cluster.go
+++ b/api/pkg/handler/v1/cluster/cluster.go
@@ -5,7 +5,6 @@ import (
 	"encoding/json"
 	"fmt"
 	"io/ioutil"
-	"math"
 	"net/http"
 	"strconv"
 	"strings"
@@ -955,13 +954,17 @@ func GetMetricsEndpoint(projectProvider provider.ProjectProvider, privilegedProj
 			availableResources[n.Name] = n.Status.Allocatable
 		}
 
-		seedAdminClient := privilegedClusterProvider.GetSeedClusterAdminRuntimeClient()
-
-		allNodeMetricsList := &v1beta1.NodeMetricsList{}
-		if err := seedAdminClient.List(ctx, allNodeMetricsList, ctrlruntimeclient.InNamespace(cluster.Status.NamespaceName)); err != nil {
+		dynamicClient, err := clusterProvider.GetAdminClientForCustomerCluster(cluster)
+		if err != nil {
 			return nil, common.KubernetesErrorToHTTPError(err)
 		}
 
+		allNodeMetricsList := &v1beta1.NodeMetricsList{}
+		if err := dynamicClient.List(ctx, allNodeMetricsList); err != nil {
+			return nil, common.KubernetesErrorToHTTPError(err)
+		}
+
+		seedAdminClient := privilegedClusterProvider.GetSeedClusterAdminRuntimeClient()
 		podMetricsList := &v1beta1.PodMetricsList{}
 		if err := seedAdminClient.List(ctx, podMetricsList, &ctrlruntimeclient.ListOptions{Namespace: fmt.Sprintf("cluster-%s", cluster.Name)}); err != nil {
 			return nil, common.KubernetesErrorToHTTPError(err)
@@ -1009,13 +1012,9 @@ func convertClusterMetrics(podMetrics *v1beta1.PodMetricsList, nodeMetrics []v1b
 		}
 	}
 	fractionCPU := float64(clusterMetrics.NodesMetrics.CPUTotalMillicores) / float64(clusterMetrics.NodesMetrics.CPUAvailableMillicores) * 100
-	if !math.IsNaN(fractionCPU) {
-		clusterMetrics.NodesMetrics.CPUUsedPercentage += int64(fractionCPU)
-	}
+	clusterMetrics.NodesMetrics.CPUUsedPercentage += int64(fractionCPU)
 	fractionMemory := float64(clusterMetrics.NodesMetrics.MemoryTotalBytes) / float64(clusterMetrics.NodesMetrics.MemoryAvailableBytes) * 100
-	if !math.IsNaN(fractionMemory) {
-		clusterMetrics.NodesMetrics.MemoryUsedPercentage += int64(fractionMemory)
-	}
+	clusterMetrics.NodesMetrics.MemoryUsedPercentage += int64(fractionMemory)
 
 	for _, podMetrics := range podMetrics.Items {
 		for _, container := range podMetrics.Containers {

--- a/api/pkg/handler/v1/cluster/cluster_test.go
+++ b/api/pkg/handler/v1/cluster/cluster_test.go
@@ -1889,11 +1889,11 @@ func TestGetClusterMetrics(t *testing.T) {
 			},
 			ExistingNodeMetrics: []*v1beta1.NodeMetrics{
 				{
-					ObjectMeta: metav1.ObjectMeta{Name: "venus", Namespace: "cluster-defClusterID"},
+					ObjectMeta: metav1.ObjectMeta{Name: "venus"},
 					Usage:      map[corev1.ResourceName]resource.Quantity{"cpu": cpuQuantity, "memory": memoryQuantity},
 				},
 				{
-					ObjectMeta: metav1.ObjectMeta{Name: "mars", Namespace: "cluster-defClusterID"},
+					ObjectMeta: metav1.ObjectMeta{Name: "mars"},
 					Usage:      map[corev1.ResourceName]resource.Quantity{"cpu": cpuQuantity, "memory": memoryQuantity},
 				},
 			},
@@ -1932,18 +1932,18 @@ func TestGetClusterMetrics(t *testing.T) {
 			},
 			ExistingNodeMetrics: []*v1beta1.NodeMetrics{
 				{
-					ObjectMeta: metav1.ObjectMeta{Name: "venus", Namespace: "cluster-defClusterID"},
+					ObjectMeta: metav1.ObjectMeta{Name: "venus"},
 					Usage:      map[corev1.ResourceName]resource.Quantity{"cpu": cpuQuantity, "memory": memoryQuantity},
 				},
 				{
-					ObjectMeta: metav1.ObjectMeta{Name: "mars", Namespace: "cluster-defClusterID"},
+					ObjectMeta: metav1.ObjectMeta{Name: "mars"},
 					Usage:      map[corev1.ResourceName]resource.Quantity{"cpu": cpuQuantity, "memory": memoryQuantity},
 				},
 			},
 		},
-		// scenario 3
+		// scenario 2
 		{
-			Name:             "scenario 3: the user John can not get Bob's cluster metrics",
+			Name:             "scenario 2: the user John can not get Bob's cluster metrics",
 			Body:             ``,
 			ExpectedResponse: `{"error":{"code":403,"message":"forbidden: \"john@acme.com\" doesn't belong to the given project = my-first-project-ID"}}`,
 			ClusterToGet:     test.GenDefaultCluster().Name,
@@ -1975,11 +1975,11 @@ func TestGetClusterMetrics(t *testing.T) {
 			},
 			ExistingNodeMetrics: []*v1beta1.NodeMetrics{
 				{
-					ObjectMeta: metav1.ObjectMeta{Name: "venus", Namespace: "cluster-defClusterID"},
+					ObjectMeta: metav1.ObjectMeta{Name: "venus"},
 					Usage:      map[corev1.ResourceName]resource.Quantity{"cpu": cpuQuantity, "memory": memoryQuantity},
 				},
 				{
-					ObjectMeta: metav1.ObjectMeta{Name: "mars", Namespace: "cluster-defClusterID"},
+					ObjectMeta: metav1.ObjectMeta{Name: "mars"},
 					Usage:      map[corev1.ResourceName]resource.Quantity{"cpu": cpuQuantity, "memory": memoryQuantity},
 				},
 			},

--- a/api/pkg/handler/v1/node/node.go
+++ b/api/pkg/handler/v1/node/node.go
@@ -482,7 +482,7 @@ func ListNodeDeploymentMetrics(projectProvider provider.ProjectProvider, privile
 	return func(ctx context.Context, request interface{}) (interface{}, error) {
 		req := request.(nodeDeploymentMetricsReq)
 		clusterProvider := ctx.Value(middleware.ClusterProviderContextKey).(provider.ClusterProvider)
-		privilegedClusterProvider := ctx.Value(middleware.PrivilegedClusterProviderContextKey).(provider.PrivilegedClusterProvider)
+
 		cluster, err := cluster.GetCluster(ctx, projectProvider, privilegedProjectProvider, userInfoGetter, req.ProjectID, req.ClusterID, nil)
 		if err != nil {
 			return nil, err
@@ -508,11 +508,14 @@ func ListNodeDeploymentMetrics(projectProvider provider.ProjectProvider, privile
 			}
 		}
 
-		seedAdminClient := privilegedClusterProvider.GetSeedClusterAdminRuntimeClient()
+		dynamicCLient, err := clusterProvider.GetAdminClientForCustomerCluster(cluster)
+		if err != nil {
+			return nil, common.KubernetesErrorToHTTPError(err)
+		}
 
 		nodeDeploymentNodesMetrics := make([]v1beta1.NodeMetrics, 0)
 		allNodeMetricsList := &v1beta1.NodeMetricsList{}
-		if err := seedAdminClient.List(ctx, allNodeMetricsList, ctrlruntimeclient.InNamespace(cluster.Status.NamespaceName)); err != nil {
+		if err := dynamicCLient.List(ctx, allNodeMetricsList); err != nil {
 			return nil, common.KubernetesErrorToHTTPError(err)
 		}
 

--- a/api/pkg/handler/v1/node/node_test.go
+++ b/api/pkg/handler/v1/node/node_test.go
@@ -1790,7 +1790,7 @@ func TestNodeDeploymentMetrics(t *testing.T) {
 			},
 			ExistingMetrics: []*v1beta1.NodeMetrics{
 				{
-					ObjectMeta: metav1.ObjectMeta{Name: "venus-1", Namespace: "cluster-defClusterID"},
+					ObjectMeta: metav1.ObjectMeta{Name: "venus-1"},
 					Usage:      map[corev1.ResourceName]resource.Quantity{"cpu": cpuQuantity, "memory": memoryQuantity},
 				},
 			},
@@ -1818,7 +1818,7 @@ func TestNodeDeploymentMetrics(t *testing.T) {
 			},
 			ExistingMetrics: []*v1beta1.NodeMetrics{
 				{
-					ObjectMeta: metav1.ObjectMeta{Name: "venus-1", Namespace: "cluster-defClusterID"},
+					ObjectMeta: metav1.ObjectMeta{Name: "venus-1"},
 					Usage:      map[corev1.ResourceName]resource.Quantity{"cpu": cpuQuantity, "memory": memoryQuantity},
 				},
 			},

